docs/proposals/20201019-prometheus-rules-in-observatorium-api.md

```diff
@@ -0,0 +1,271 @@
+
+While this may impose some file splitting costs on the user's side, we avoid ambiguity such as valid questions such as:
+* Did all of my rules get stored or just some?
+* Are my alerting rules being evaluated or just the recording rules?
+
+Finally, warnings are easy to ignore especially if users are using automated process to interact with the API.
+
+#### Storage layer
+
+We will use a persistent storage mechanism that is external to the API deployment (i.e. no local storage). This enables us to satisfy our requirement of horizontally scaling the rule storage backend.
+
+While we have explicitly de-coupled the rule storage backend from the API, we require a concrete implementaion.
+
+We have a number of options for the storage layer backing the rules storage backend service.
+
+##### 1. ETCD
+
+One option would be to use the Kubernetes' control plane as the backing data store for tenant's recording rules.
+
+Pros:
+* Lowest possible overhead. No external storage is required to be orchestrated.
+* ETCD data model well suited for recording rules i.e. YAML.
+* Storage permissions are managed as Kubernetes objects.
+* We could leverage Recording Rules CRDs and get validation for free (?)
+
+Cons:
+* Implicitly ties Observatorium's implementation into a Kubernetes environment. Hard to back out.
+* Large blast radius. With a large number of tenants and recording rules, we could un-intentionally impact the Kubernetes control plane's performance.
+* ETCD has a limit of 1.5MB per key / value pair so we can't store large files (but Prometheus operator seems to do this ok?).
+
+##### 2. Object Storage
+
+Pros:
+* Users of Thanos project are likely to already have object storage configured and ready to use.
+* Recording rule data model is a good fit for object storage i.e. YAML.
+
+Cons:
+* Unclear consistency semantics depending on object storage provider i.e. in the case of lots of writes, who wins?
+* Requires tenants to manage another set of secret credentials.
+
+##### 3. RDBMS
+
+Another option we have considered is to provision and use a relational database i.e. Postgres.
+
+Pros:
+* Strong isolation and update semantics.
+* Enables us to use multi-dimensional
+
+Cons:
+* We don't currently run RDBMS, and don't have any experience doing so.
+* Significant infrastructure overhead - tenants required to BYO database.
+* Requires tenants to manage another set of secret credentials.
+
+##### Storage Decision
+
+This was discussed in the [Observatorium Community Meeting](https://docs.google.com/document/d/1jLvOH0Lllt-ShXVgWeeHBGPb3ZuagirabD_hbq41EKs/edit#heading=h.wxenjcrf4iee) on 2021-07-31.
+
+The consensus among the group was that object storage offered us the best combination of ease of use, familiarity for existing users, and also compatability with the required operations against tenants rules.
```

Comment:
Nice! I like this idea, BUT I would also consider pros / cons I commented in. I think we missed important part where object storage is hidden by our API so we can change storage later and iterate

Reply:
Actually your diagram has it all, just above items might be not clear about those